libcaf_net/caf/net/octet_stream/transport.cpp

```diff
@@ -12,361 +12,478 @@
         return;
-      default:
-        return fail(make_error(sec::socket_operation_failed));
+      } else if (consumed == 0) {
+        if (next_) {
+          // When switching protocol, the new layer has never seen the data, so
+          // we might just re-invoke the same data again.
+          if (!switch_to_next_protocol())
+            return;
+        } else {
+          // See whether the next iteration would change what we pass to the
+          // application (max_read_size_ may have changed). Otherwise, we'll try
+          // again later.
+          delta_offset_ = static_cast<ptrdiff_t>(n);
+          if (n == std::min(buffered_, size_t{max_read_size_}))
+            return;
+        }
+      } else {
+        if (next_ && !switch_to_next_protocol())
+          return;
+        // Shove the unread bytes to the beginning of the buffer and continue
+        // to the next loop iteration.
+        auto prev = static_cast<ptrdiff_t>(buffered_);
+        buffered_ -= static_cast<size_t>(consumed);
+        delta_offset_ = 0;
+        if (buffered_ > 0)
+          std::copy(read_buf_.begin() + consumed, read_buf_.begin() + prev,
+                    read_buf_.begin());
+      }
     }
-  } else {
-    // write() returns 0 if the connection was closed.
-    return fail(make_error(sec::socket_disconnected));
   }
-}
 
-void transport::abort(const error& reason) {
-  up_->abort(reason);
-  flags_.shutting_down = true;
+  /// Calls abort on the upper layer and deregisters the transport from events.
+  void fail(const error& reason) {
+    auto lg = log::net::trace("reason = {}", reason);
+    up_->abort(reason);
+    up_.reset();
+    parent_->deregister();
+    parent_->shutdown();
+  }
+
+  // -- member variables -------------------------------------------------------
+
+  /// Stores temporary flags.
+  flags_t flags_;
+
+  /// Caches the config parameter for limiting max. socket operations.
+  size_t max_consecutive_reads_ = defaults::middleman::max_consecutive_reads;
+
+  /// Caches the write buffer size of the socket.
+  uint32_t max_write_buf_size_ = 0;
+
+  /// Stores what the user has configured as read threshold.
+  uint32_t min_read_size_ = 0;
+
+  /// Stores what the user has configured as max. number of bytes to receive.
+  uint32_t max_read_size_ = 0;
+
+  /// Stores how many bytes are currently buffered, i.e., how many bytes from
+  /// `read_buf_` are filled with actual data.
+  size_t buffered_ = 0;
+
+  /// Stores the offset in `read_buf_` since last calling
+  /// `upper_layer_.consume`.
+  size_t delta_offset_ = 0;
+
+  /// Caches incoming data.
+  byte_buffer read_buf_;
+
+  /// Caches outgoing data.
+  byte_buffer write_buf_;
+
+  /// Processes incoming data and generates outgoing data.
+  upper_layer_ptr up_;
+
+  /// Points to the associated socket manager below.
+  socket_manager* parent_ = nullptr;
+
+  /// Configures how we read and write to the socket.
+  std::unique_ptr<policy> policy_ = nullptr;
```

Comment:
That's just what the default constructor will do anyways.

```suggestion
  std::unique_ptr<policy> policy_;
```